### PR TITLE
Tracking converter for ROS2 Foxy

### PR DESCRIPTION
**Pull request by daniqsilva25**
## Overview
Author: [@daniqsilva25](https://github.com/daniqsilva25)

## Issue 
Issue link (if present): https://github.com/luxonis/depthai-ros/issues/146
Issue description: Absence of a tracking converter. This PR adds a tracking converter to enable object tracking with ROS.
Related PRs

## Changes
ROS distro: Foxy
List of changes:
* Added other spatial detection converter that uses standard ROS2 vision msgs for more compatibility with other ROS2 code
* Added a track detection converter and track spatial detection converter for tracking purposes with 2D or 3D detections, respectively
* Added new ROS2 messages for tracking: TrackDetection2D and TrackDetection2DArray. The first message adds two important fields for tracking tasks: tracking age and tracking status, both available in the depthai pipeline. The second message is just an array of the first one.

## Testing
Hardware used: OAK-1, OAK-D
Depthai library version: 2.21.2.0

## Visuals from testing
Add screenshots/gifs/videos from RVIZ or other visualizers demonstrating the effect of the changes when applicable.

### CODE REVIEW

**Comment by daniqsilva25** [comment]
Hi @Serafadam @saching13, I changed the indentation format of the source files from tabs to spaces after seeing the output of the clang-format test. Can you please review this? Thanks!

File: depthai_bridge/include/depthai_bridge/TrackDetectionConverter.hpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+#include <deque>
+#include <memory>
+#include <string>
+#include <vision_msgs/msg/detection2_d_array.hpp>
```
Comment (Serafadam) [inline depthai_bridge/include/depthai_bridge/TrackDetectionConverter.hpp]: Currently we are following guideline to keep `<>` tags for system libraries, could you change those includes to use quotes?

File: depthai_bridge/include/depthai_bridge/TrackDetectionConverter.hpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    int _width, _height;
+    const std::string _frameName;
+    bool _normalized;
+    float _thresh;
+    std::chrono::time_point<std::chrono::steady_clock> _steadyBaseTime;
+    rclcpp::Time _rosBaseTime;
+    bool _getBaseDeviceTimestamp;
+};
+
+/** TODO(sachin): Do we need to have ros msg -> dai bounding box ?
```
Comment (Serafadam) [inline depthai_bridge/include/depthai_bridge/TrackDetectionConverter.hpp]: I guess this comment is a leftover from copying from other converters, can you remove it?

File: depthai_bridge/include/depthai_bridge/SpatialDetection2DConverter.hpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+#include "depthai/pipeline/datatype/SpatialImgDetections.hpp"
+
+namespace dai {
+
+namespace ros {
+
+namespace VisionMsgs = vision_msgs::msg;
+using SpatialDetection2DArrayPtr = VisionMsgs::Detection2DArray;
+
+class SpatialDetection2DConverter {
```
Comment (Serafadam) [inline depthai_bridge/include/depthai_bridge/SpatialDetection2DConverter.hpp:18]: What is the exact use case here? Current SpatialDetectionConverter has [a method](https://github.com/luxonis/depthai-ros/blob/humble/depthai_bridge/src/SpatialDetectionConverter.cpp#L95) to output vision_msgs format

File: depthai_bridge/src/TrackDetectionConverter.cpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        opDetectionMsg.detections[i].bbox.size_x = xSize;
+        opDetectionMsg.detections[i].bbox.size_y = ySize;
+        opDetectionMsg.detections[i].is_tracking = true;
+        std::stringstream track_id_str;
+        track_id_str << "" << t.id;
+        opDetectionMsg.detections[i].tracking_id = track_id_str.str();
+        opDetectionMsg.detections[i].tracking_age = t.age;
+        opDetectionMsg.detections[i].tracking_status = static_cast<int32_t>(t.status);
+
+        // converting mm to meters since per ros rep-103 lenght should always be in meters
```
Comment (Serafadam) [inline depthai_bridge/src/TrackDetectionConverter.cpp]: Why is that commented out part left here?

File: depthai_bridge/include/depthai_bridge/TrackSpatialDetectionConverter.hpp [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    int _width, _height;
+    const std::string _frameName;
+    bool _normalized;
+    float _thresh;
+    std::chrono::time_point<std::chrono::steady_clock> _steadyBaseTime;
+    rclcpp::Time _rosBaseTime;
+    bool _getBaseDeviceTimestamp;
+};
+
+/** TODO(sachin): Do we need to have ros msg -> dai bounding box ?
```
Comment (Serafadam) [inline depthai_bridge/include/depthai_bridge/TrackSpatialDetectionConverter.hpp]: Same situation with comment

File: depthai_bridge/src/TrackDetectionConverter.cpp [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+TrackDetection2DArrayPtr TrackDetectionConverter::toRosMsgPtr(
+    std::shared_ptr<dai::Tracklets> trackData) {
+
+    std::deque<DepthaiMsgs::TrackDetection2DArray> msgQueue;
+    toRosMsg(trackData, msgQueue);
+    auto msg = msgQueue.front();
+
+#ifdef IS_ROS2
```
Comment (Serafadam) [inline depthai_bridge/src/TrackDetectionConverter.cpp]: These checks can be removed as we currently have separate branches for ROS1&2

File: depthai_bridge/src/TrackSpatialDetectionConverter.cpp [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+TrackDetection2DArrayPtr TrackSpatialDetectionConverter::toRosMsgPtr(
+    std::shared_ptr<dai::Tracklets> trackData) {
+
+    std::deque<DepthaiMsgs::TrackDetection2DArray> msgQueue;
+    toRosMsg(trackData, msgQueue);
+    auto msg = msgQueue.front();
+
+#ifdef IS_ROS2
```
Comment (Serafadam) [inline depthai_bridge/src/TrackSpatialDetectionConverter.cpp]: Same as in the other converter

File: depthai_bridge/src/TrackDetectionConverter.cpp [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        yMax = roi.bottomRight().y;
+
+        float xSize = xMax - xMin;
+        float ySize = yMax - yMin;
+        float xCenter = xMin + xSize / 2.;
+        float yCenter = yMin + ySize / 2.;
+
+        opDetectionMsg.detections[i].results.resize(1);
+
+#if defined(IS_GALACTIC) || defined(IS_HUMBLE)
```
Comment (Serafadam) [inline depthai_bridge/src/TrackDetectionConverter.cpp]: Those checks can be removed as we have separate branches for different distros now

File: depthai_bridge/src/TrackSpatialDetectionConverter.cpp [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        yMax = roi.bottomRight().y;
+
+        float xSize = xMax - xMin;
+        float ySize = yMax - yMin;
+        float xCenter = xMin + xSize / 2.;
+        float yCenter = yMin + ySize / 2.;
+
+        opDetectionMsg.detections[i].results.resize(1);
+
+#if defined(IS_GALACTIC) || defined(IS_HUMBLE)
```
Comment (Serafadam) [inline depthai_bridge/src/TrackSpatialDetectionConverter.cpp]: Same here with checks

**Review by Serafadam — CHANGES_REQUESTED** [review]
Hi, thanks very much for the PR, we really appreciate it! I have left some comments in the code, but I also wanted to ask few questions:
1. Could you add an example node in `depthai_examples` using those new converters?
2. Would it be possible to also create a PR for `humble` branch? As foxy is EOL, those converters won't be available in new Docker images/deb releases.

File: depthai_bridge/include/depthai_bridge/TrackSpatialDetectionConverter.hpp (hunk 5 above)
Comment (daniqsilva25) [reply depthai_bridge/include/depthai_bridge/TrackSpatialDetectionConverter.hpp]: I've removed the comment.

File: depthai_bridge/src/TrackDetectionConverter.cpp (hunk 8 above)
Comment (daniqsilva25) [reply depthai_bridge/src/TrackDetectionConverter.cpp]: Ok, I've removed the checks.

File: depthai_bridge/src/TrackDetectionConverter.cpp (hunk 4 above)
Comment (daniqsilva25) [reply depthai_bridge/src/TrackDetectionConverter.cpp]: For no reason. I've removed it.

File: depthai_bridge/src/TrackDetectionConverter.cpp (hunk 6 above)
Comment (daniqsilva25) [reply depthai_bridge/src/TrackDetectionConverter.cpp]: Yes. Done!

File: depthai_bridge/src/TrackSpatialDetectionConverter.cpp (hunk 9 above)
Comment (daniqsilva25) [reply depthai_bridge/src/TrackSpatialDetectionConverter.cpp]: Yes, done.

File: depthai_bridge/src/TrackSpatialDetectionConverter.cpp (hunk 7 above)
Comment (daniqsilva25) [reply depthai_bridge/src/TrackSpatialDetectionConverter.cpp]: Done.

File: depthai_bridge/include/depthai_bridge/TrackDetectionConverter.hpp (hunk 1 above)
Comment (daniqsilva25) [reply depthai_bridge/include/depthai_bridge/TrackDetectionConverter.hpp]: Ok, I've changed it.

File: depthai_bridge/include/depthai_bridge/TrackDetectionConverter.hpp (hunk 2 above)
Comment (daniqsilva25) [reply depthai_bridge/include/depthai_bridge/TrackDetectionConverter.hpp]: Yes, I've removed it.

File: depthai_bridge/include/depthai_bridge/SpatialDetection2DConverter.hpp (hunk 3 above)
Comment (daniqsilva25) [reply depthai_bridge/include/depthai_bridge/SpatialDetection2DConverter.hpp:18]: Yes, that is true, but I think Detection3DArray is not necessary, Detection2DArray is enough. And actually you are not making use of it [here](https://github.com/luxonis/depthai-ros/blob/10930361f8d44725b656a1c0b33f5e2975f550b6/depthai_bridge/src/SpatialDetectionConverter.cpp#L139).
But maybe I can add another function to the exitent SpatialDetectionConverter to publish Detection2DArray, let me know your opinion on this.

Comment (daniqsilva25) [reply depthai_bridge/include/depthai_bridge/SpatialDetection2DConverter.hpp:18]: @Serafadam what should I do in this case?

Comment (Serafadam) [reply depthai_bridge/include/depthai_bridge/SpatialDetection2DConverter.hpp:18]: Hi, if it really needed then I would update spatial detection converter to have an option to choose which detection array to publish, with default being detection3d array.

**Comment by Serafadam** [comment]
@daniqsilva25 Could you respond to those points?
- Could you add an example node in depthai_examples using those new converters?
- Would it be possible to also create a PR for humble branch? As foxy is EOL, those converters won't be available in new Docker images/deb releases.
Also, it seems parts of the code need reformatting.

**Comment by daniqsilva25** [comment]
@Serafadam I created the PR for the humble branch [#505](https://github.com/luxonis/depthai-ros/pull/505) and added some examples in both branches (foxy and humble).

**Comment by daniqsilva25** [comment]
@Serafadam I've reformatted the code of the converters.

**Comment by Serafadam** [comment]
Hi @daniqsilva25 , overall LGTM, but it seems that it still fails formatting checks,with clang format you can do the reformatting by running in your workspace `ament_clang_format src/depthai-ros --config ${workspaceFolder}/src/depthai-ros/.clang-format --reformat`

**Comment by daniqsilva25** [comment]
> Hi @daniqsilva25 , overall LGTM, but it seems that it still fails formatting checks,with clang format you can do the reformatting by running in your workspace `ament_clang_format src/depthai-ros --config ${workspaceFolder}/src/depthai-ros/.clang-format --reformat`

Hi @Serafadam, just did that for both PRs.
Thanks.

**Pull request merged**